[Form][DX] FileType "multiple" fixes

yceruto opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/symfony/symfony/issues/12547
| License       | MIT
| Doc PR        | -

# (1st) Derive "data_class" option from passed "multiple" option

Information
-------------

Following this tutorial ["How to Upload Files"][1] but storing many `brochures` instead of one, i.e.:

```php
// src/AppBundle/Entity/Product.php

class Product 
{
    /**
     * @var string[]
     *
     * @ORM\Column(type="array")
     */
    private $brochures;
    
    //...
}
```

```php
//src/AppBundle/Form/ProductType.php

$builder->add('brochures', FileType::class, array(
    'label' => 'Brochures (PDF files)',
    'multiple' => true,
));
```

The Problem
--------------

I found a pain point here when the form is loaded again after save some brochures (Exception):

> The form's view data is expected to be an instance of class Symfony\Component\HttpFoundation\File\File, but is a(n) array. You can avoid this error by setting the "data_class" option to null or by adding a view transformer that transforms a(n) array to an instance of Symfony\Component\HttpFoundation\File\File.

The message is very clear, but counter-intuitive in this case, because the form field (`FileType`) was configured with `multiple = true`, so IMHO it shouldn't expect a `File` instance but an array of them at all events.

The PR's effect
---------------

**Before:**

```php
$form = $this->createFormBuilder($product)
    ->add('brochures', FileType::class, [
        'multiple' => true,
	'data_class' => null, // <---- mandatory
    ])
    ->getForm();
```

**After:**

```php
$form = $this->createFormBuilder($product)
    ->add('brochures', FileType::class, [
        'multiple' => true,
    ])
    ->getForm();
```

# (2nd) Return empty `array()` at submit no file

Information
-------------

Based on the same information before, but adding some constraints:

```php
// src/AppBundle/Entity/Product.php

use Symfony\Component\Validator\Constraints as Assert;

class Product 
{
    /**
     * @var string[]
     *
     * @ORM\Column(type="array")
     *
     * @Assert\Count(min="1") // or @Assert\NotBlank()
     * @Assert\All({
     *     @Assert\File(mimeTypes = {"application/pdf", "application/x-pdf"})
     * })
     * 
     */
    private $brochures;
}
```

This should require at least one file to be stored.

The Problem
--------------

But, when no file is uploaded at submit the form, it's valid completely. The submitted data for this field was `array(null)` so the constraints pass without any problem:

* `@Assert\Count(min="1")` pass! because contains at least one element (No matter what)
* `@Assert\NotBlank()` it could pass! because no `false` and no `empty()`
* `@Assert\File()` pass! because the element is `null`

Apart from that really we expecting an empty array instead.

The PR's effect
----------------

**Before:**

```php
// src/AppBundle/Entity/Product.php

use Symfony\Component\Validator\Constraints as Assert;

class Product 
{
    /**
     * @var string[]
     *
     * @ORM\Column(type="array")
     *
     * @Assert\All({
     *     @Assert\NotBlank,
     *     @Assert\File(mimeTypes = {"application/pdf", "application/x-pdf"})
     * })
     * 
     */
    private $brochures;
}
```

**After:**

```php
// src/AppBundle/Entity/Product.php

use Symfony\Component\Validator\Constraints as Assert;

class Product 
{
    /**
     * @var string[]
     *
     * @ORM\Column(type="array")
     *
     * @Assert\Count(min="1") // or @Assert\NotBlank
     * @Assert\All({
     *     @Assert\File(mimeTypes = {"application/pdf", "application/x-pdf"})
     * })
     * 
     */
    private $brochures;
}
```


  [1]: http://symfony.com/doc/current/controller/upload_file.html

yceruto committed:
[Form] FileType "multiple" fixes

yceruto commented:
Failure unrelated

ro0NL commented:
Should we also return an empty array, if multiple? ie. like the choice type does: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php#L244

yceruto commented:
@ro0NL yeah, I had already planned it in a separated PR :+1:

ro0NL commented:
Could be one PR if you ask me :)

yceruto commented:
I've updated this PR with the 2nd FileType improvement, as it depends on the 1st.

ogizanagi commented:
@yceruto : This transformer looks strange to me.

> Edit: because empty_data doesn't work when no file is uploaded (with multiple = true it submit array(null) instead of null or array()) so we need a transformer to fix that and validate also the array of File and UploadedFile .

Would it be feasible to listen on `FormEvents::PRE_SUBMIT` and set an empty array if the event data is `array(null)` ?

yceruto commented:
@ogizanagi:

> Would it be feasible to listen on FormEvents::PRE_SUBMIT and set an empty array (or even $form->getConfig()->getEmptyData() result) if the event data is array(null) ?

Yes! could be (my first thought was that) but, what about the data flow validation? Do you think that the transformer should be removed completely (even when there is no bijective transformation)?

ogizanagi commented:
@yceruto : I think it should be removed completely. It's not its responsibility to validate the data, except when it prevents transformation. There is no transformation here.

ogizanagi commented on src/Symfony/Component/Form/Tests/Extension/Core/Type/FileTypeTest.php:
```diff
@@ -44,6 +44,34 @@ public function testSubmitEmpty()
+        ))->getForm();
+
+        // submitted data when an input file is uploaded
+        // without choice any file
```
```diff
- without choice any file
+ without choosing any file
```

?

yceruto commented:
We need be consistent with FileType data flow, otherwise when `multiple = true` any data type can be passed to this one (even when it do nothing with these values).

yceruto commented:
I've updated the implementation:

* DataTransformer removed
* Added `PRE_SUBMIT` listener to change `array(null)` to default empty data.
* Added empty data default value (`array()`) for `'multiple' => true`

ping @ogizanagi

ogizanagi commented on src/Symfony/Component/Form/Extension/Core/Type/FileType.php:
```diff
@@ -12,15 +12,34 @@
+            $builder->addEventListener(FormEvents::PRE_SUBMIT, function (FormEvent $event) {
+                // submitted data for an input file (no required) without choosing any file
+                if (array(null) === $event->getData()) {
+                    $event->setData($event->getForm()->getConfig()->getEmptyData());
```
Beware, the empty data could be a callable. See https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php#L96-L100

yceruto replied:
Opps! updated, added support to callable `empty_data`

ogizanagi reviewed (approved):
👍 

Status: Reviewed

ro0NL commented on src/Symfony/Component/Form/Extension/Core/Type/FileType.php:
```diff
@@ -12,15 +12,40 @@
+                $data = $event->getData();
+
+                // submitted data for an input file (no required) without choosing any file
+                if (array(null) === $data) {
```
Can we expect multiple nulls here? `array(null, null, ...)`?

yceruto replied:
Nope! `array(null)` mean no file uploaded, otherwise we expect array of UploadedFile instances.

ogizanagi commented:
Status: Reviewed

(bis 😆 )

HeahDude commented:
👍

Nice job with this PR @yceruto! What about having this merged in 2.7 as a bug fix instead?

yceruto commented:
For me will be great to add this one since 2.7 (Rebasing...)

Status: Needs Work

yceruto commented:
Rebased to 2.7 finished :sweat_smile:

Status: Needs Review

HeahDude commented on src/Symfony/Component/Form/Extension/Core/Type/FileType.php:
```diff
@@ -12,15 +12,40 @@
+                $form = $event->getForm();
+                $data = $event->getData();
+
+                // submitted data for an input file (no required) without choosing any file
```
typo `not required`

HeahDude commented on src/Symfony/Component/Form/Tests/Extension/Core/Type/FileTypeTest.php:
```diff
@@ -44,6 +44,33 @@ public function testSubmitEmpty()
+            'multiple' => true,
+        ))->getForm();
+
+        // submitted data when an input file is uploaded without choosing any file
```
does it deserve a test case without default required?

yceruto replied:
why not?

yceruto replied:
This is often the case, for example, when editing a form requesting to change the related image optionally.

yceruto committed:
fixed typo

yceruto commented:
Can anyone remove ``Feature`` and add ``Bug`` label ?

yceruto commented:
It's ready for review again now in 2.7 branch.

Thank you @ogizanagi for you previous review.

ro0NL commented on src/Symfony/Component/Form/Extension/Core/Type/FileType.php:
```diff
@@ -12,15 +12,40 @@
+                $data = $event->getData();
+
+                // submitted data for an input file (not required) without choosing any file
+                if (array(null) === $data) {
```
Still not sure about this, by default only 1 `<input>` is rendered, but it can in fact be many (ie. created by the client with a "Add file" button), hence `multiple`.

So if many inputs are submitted, and only few files are uploaded, we get `array(null, <file>, null, null)` for instance. And as the data model is a collection, i guess `array(<file>)` is actually expected.

ogizanagi replied:
What you describe here looks more like a `CollectionType` of `FileType` than a single `FileType` with `multiple` option set to `true` (which is rendered with a single `<input type="file" multiple>`), right ?

ro0NL replied:
I guess im not sure this involves / should involve the collection type, the `multiple` option only works with it?

What if multiple inputs are there somehow.. we currently get on submit

![image](https://cloud.githubusercontent.com/assets/1047696/20541616/53b1ea64-b0fe-11e6-8d06-d326e5fceba8.png)

Imo. we should filter nulls, or we shouldnt (not only a single null). Maybe make it an option?

ro0NL replied:
Yeah, but that makes the `multiple` option pretty much useless, as the collection type doesnt need it. Imo. this PR fixes the filetype when used standalone _with_ the multiple option..

ro0NL replied:
And yeah, people could actually rely on this format.. so i think it should be opt-in, or target master otherwise.

ogizanagi replied:
> Yeah, but that makes the multiple option pretty much useless, as the collection type doesnt need it. Imo. this PR fixes the filetype when used standalone with the multiple option..

I'm a bit lost with your comment... I never told you to use a collection of `FileType` along with the multiple option...
I'm talking about a **single** `FileType` **with** the `multiple` set to true, which is indeed what is trying to solve this PR.

I don't follow you, but I hope @yceruto or anyone else will better than me 😄

ro0NL replied:
Just saying, technically expecting an array with files or nulls is correct for the filetype with multiple set to true.

Symfony only rendering 1 doesnt sound multiple to me.. so, in fact, we should expect many (again; for being _multiple_ and not being dependent on the template implementation).

Hope you follow :)

ogizanagi replied:
So, what about using a `TextType` with a `multiple` option and expect it to properly map multiple `<input type="text">` and an array of strings?..

I'm just saying it's not how this form type is meant to be used. That's why the `CollectionType` one exists. It does not withdraw the legitimacy for the `FileType` to have a `multiple` option, as its html counterpart propose it.

Alright, let's move forwards, what would you suggest? Something like:

```diff
- if (array(null) === $data) {
+ if (array() === array_filter($data)) {
```

to detect empty data?

ro0NL replied:
Yes.

To me [this](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Core/Type/FileType.php#L27) affects the underlying data format. The text type has no multiple option; doing `texttype[x]` should indeed be done with a collection type.

The multiple option is specific to the file type, giving you `filetype[]` without the collection type.

ro0NL replied:
@ogizanagi think i get your point, i actually didnt consider the _HTML_ multiple attribute on this.. (`<input name="foo[]" type="iile" multiple>`). Which indeed limits the values to `array(null)` or `array(File, File, etc.)`.

:+1: for the current approach, my bad.

yceruto replied:
@ro0NL : There is no way to submit `array(null, <file>, null, null)` for a single `FileType` with `multiple` set to `true`. You have just three options: 

Select `<input type="file" multiple="true">`:
1- one file (`array(<UploadedFile>)`)
2- many files (`array(<UploadedFile>, <UploadedFile>, ...)`)
3- none file (`array(null)`)

This is about the 3rd option and normalize the empty data to `array()` when this happen.

> So if many inputs are submitted, and only few files are uploaded, we get `array(null, <file>, null, null)` for instance. And as the data model is a collection, i guess `array(<file>)` is actually expected.

I guess you are talking about a `CollectionType` with `entry_type` set to `FileType::class` (not multiple) ? anyway: `array(null, <file>, null, null)` each of them is assigned to a  `FileType`.

ro0NL replied:
See my screenshot; the data can, _in fact_, be manipulated by manipulating the DOM. Which can even be considered a security issue ;-) if we advertise with your 3 options.

ro0NL replied:
That still doesnt guarantee an array with _only_ files... which seems bad DX right? I prefer your first patch :+1:

ogizanagi replied:
The patches are pretty equivalent, but the second ones simply add the `array(null, <file>, null, null)` to `array(<file>)` normalization (and avoid rejecting a file name "0").

ro0NL replied:
Correct, i think it should set  `setData(array_filter($data))` if not empty, additionally.

ogizanagi replied:
It does 😅

ro0NL replied:
I see :) im going to sleep now. But yeah the 2nd patch is good :+1: 

@yceruto agree?

yceruto replied:
> the data can, in fact, be manipulated by manipulating the DOM or doing a raw request.

[Expected data] Only those with Data Transformer take care of this. To be honest, I'm not sure if we need cover any "weird" case/usage in this form type for now and others like `TextType`, `UrlType`, `PasswordType`, etc.

So :-1: for now :)

yceruto replied:
However, I'm in favor to validate the expected data (in/out) for those form type without Data Transformer. But as my friend @ogizanagi says, this should only be done if there is a data transformation.

ro0NL replied:
Not sure it needs a data transformer, as this is only about filtering out null values. Again; to enforce the `multiple="true"` effect server-side. It's an edge case.

This does not apply in any way to other types. Doing weird things with textypes, ie. changing  `name="texttype` to `name="texttype[weird]"`, will make it crash already (a good thing);

```
Expected argument of type "string", "array" given
```

yceruto replied:
> this is only about filtering out null values... to enforce the multiple="true" effect server-side. It's an edge case.

Imho it's already cover, we can avoid that with `All()` constraint (See in description):

```php
     * @Assert\All({
     *     @Assert\File(mimeTypes = {"application/pdf", "application/x-pdf"})
     * })
     */
    private $brochures;
```

Also by manipulating the DOM or doing a raw request we can garantee a `array()` result with `Type("array")` constraint if submit `null` value with `multiple="true"`.

ro0NL replied:
I guess we're letting the user deal with it then. Just saying, not everybody probably expects this behavior (leaving a hole in your architecture).

But we'er good then.

yceruto replied:
> not everybody probably expects this behavior.

Exactly, the form type only should cover the normal expected behavior.

> leaving a hole in your architecture

AFAIK this hole is responsibility of constraints/validators.

ro0NL replied:
> Exactly, the form type only should cover the normal expected behavior.

Imo. it should define consistent behavior :)

> AFAIK this hole is responsibility of constraints/validators.

Just saying we could avoid the whole thing by filtering out nulls out-of-the-box, having consistent behavior :) But im fine with keeping it in user-land as well.

The point is clear, thanks for letting me clarify!

if everbody's in favor with this approach, im as well :+1:

yceruto commented:
@HeahDude now in 2.7 base branch as "bugfix" any thought about this one?

It's ready for @symfony/deciders?

HeahDude commented:
👍

xabbuh commented:
:+1:

Status: Reviewed

fabpot commented:
Thank you @yceruto.
